File: jar/src/main/java/jline/api/pfqn/nc/Pfqn_propfair.java
/**
 * @file Proportionally fair allocation approximation for normalizing constants
 *
 * @since LINE 3.0
 */
package jline.api.pfqn.nc;

import de.xypron.jcobyla.Calcfc;
import de.xypron.jcobyla.Cobyla;
import jline.io.Ret;
import jline.GlobalConstants;
import jline.util.Maths;
import jline.util.matrix.Matrix;
import org.apache.commons.math3.util.FastMath;

public final class Pfqn_propfair {
    private Pfqn_propfair() {}

    /**
     * Compute the proportionally fair allocation approximation.
     *
     * @param L demands at all stations
     * @param N number of jobs for each class
     * @param Z think time for each class
     * @return normalizing constant, its logarithm, and performance metrics
     */
    public static Ret.pfqnNcXQ pfqn_propfair(Matrix L, Matrix N, final Matrix Z) {
        final int M = L.getNumRows();
        final int R = L.getNumCols();

        final double[] Nvec = N.toArray1D();
        final double[] Zvec = Z.toArray1D();
        final double[][] Lmat = L.toArray2D();

        Calcfc objFun = new Calcfc() {
            @Override
            public double compute(int n, int m, double[] x, double[] con) {
                double obj = 0.0;
                for (int r = 0; r < R; r++) {
                    obj += ((Nvec[r] - x[r] * Zvec[r]) * FastMath.log(FastMath.abs(x[r]) + GlobalConstants.FineTol));
                }
                for (int i = 0; i < M; i++) {
                    con[i] = 1.0;
                    for (int r = 0; r < R; r++) {
                        con[i] -= Lmat[i][r] * x[r];
                    }
                }
                System.arraycopy(x, 0, con, M, R);
                return -obj;
            }
        };

        double[] Xasy = new double[R];
        Cobyla.findMinimum(objFun, R, M + R, Xasy, 1.0, 1.0e-8, 0, 10000); // iprint=0: silent (no COBYLA iteration trace)

        // Z.get(0,r) IS CORRECT HERE and a `Z.get(0,` sweep should not flag it:
        // the reference indexes Z as a VECTOR too (pfqn_propfair.m:44-48 write
        // Z(r)), so it is no more tolerant of a per-delay-node Z than this is.
        // Summing the rows would MANUFACTURE a divergence rather than remove
        // one. The defect this pattern signals is taking Z straight from
        // snGetProductFormParams with no sumCols() in between; propfair has no
        // such caller.
        double lG = 0.0;
        for (int r = 0; r < R; r++) {
            double x = Xasy[r];
            lG += ((N.get(0, r) - x * Z.get(0, r)) * FastMath.log(1.0 / (x + GlobalConstants.FineTol)));
        }
        for (int r = 0; r < R; r++) {
            lG -= FastMath.log(Maths.fact(Xasy[r] * Z.get(0, r)));
        }

        double G = FastMath.exp(lG);

        Matrix Xa = new Matrix(Xasy);
        Matrix Qa = Xa.copy();
        Qa.fill(Double.NaN);

        return new Ret.pfqnNcXQ(G, lG, Xa, Qa, "propfair");
    }
}

File: jar/src/main/kotlin/jline/api/pfqn/nc/Pfqn_propfair.kt
/**
 * @file Proportionally fair allocation approximation for normalizing constants
 * 
 * Implements the proportionally fair allocation method using convex optimization
 * for computing normalizing constants in closed queueing networks. Based on Schweitzer's
 * approach and Walton's proportional fairness theory for multi-class networks.
 * 
 * @since LINE 3.0
 */
package jline.api.pfqn.nc

import de.xypron.jcobyla.Calcfc
import de.xypron.jcobyla.Cobyla
import jline.io.Ret
import jline.GlobalConstants
import jline.util.Maths
import jline.util.matrix.Matrix
import org.apache.commons.math3.util.FastMath

/**
 * Compute the proportionally fair allocation approximation
 *
 * Estimate the normalizing constant using a convex optimization program
 * that is asymptotically exact in models with single-server PS queues only.
 * The underlying optimization program is convex.
 * The script implements a heuristic to estimate the solution in the
 * presence of delay stations.
 *
 * References:
 * Schweitzer, P. J. (1979). Approximate analysis of multiclass closed networks of
 * queues. In Proceedings of the International Conference on Stochastic Control
 * and Optimization. Free Univ., Amsterdam.
 *
 * Walton, Proportional fairness and its relationship with multi-class
 * queueing networks, 2009.
 *
 * @param L - demands at all stations
 * @param N - number of jobs for each class
 * @param Z - think time for each class
 * @return normalizing constant, its logarithm, and performance metrics
 */
fun pfqn_propfair(L: Matrix, N: Matrix, Z: Matrix): Ret.pfqnNcXQ {
    val M = L.numRows
    val R = L.numCols

    val Nvec = N.toArray1D()
    val Zvec = Z.toArray1D()
    val Lmat = L.toArray2D()

    /* ---------- objective & constraints for COBYLA ---------------- */
    val objFun = Calcfc { n, m, x, con ->
        var obj = 0.0
        for (r in 0..<R) {
            obj += ((Nvec[r] - x[r] * Zvec[r]) * FastMath.log(FastMath.abs(x[r]) + GlobalConstants.FineTol))
        }
        for (i in 0..<M) {
            con[i] = 1.0
            for (r in 0..<R) {
                con[i] -= Lmat[i][r] * x[r]
            }
        }
        System.arraycopy(x, 0, con, M, R) // x ≥ 0
        -obj // maximise obj
    }

    val Xasy = DoubleArray(R)
    Cobyla.findMinimum(objFun, R, M + R, Xasy, 1.0, 1.0e-8, 2, 10000)

    /* -------------- updated lG -------------- */
    var lG = 0.0
    for (r in 0..<R) {
        val x = Xasy[r]
        lG += ((N[0, r] - x * Z[0, r]) * FastMath.log(1.0 / (x + GlobalConstants.FineTol)))
    }
    for (r in 0..<R) {
        lG -= FastMath.log(Maths.fact(Xasy[r] * Z[0, r]))
    }

    val G = FastMath.exp(lG)

    val Xa = Matrix(Xasy)
    val Qa = Xa.copy()
    Qa.fill(Double.NaN)

    return Ret.pfqnNcXQ(G, lG, Xa, Qa, "propfair")
}
/**
 * PFQN propfair algorithms
 */
@Suppress("unused")
class PfqnPropfairAlgo {
    companion object {
        // Class documentation marker for Dokka
    }
}